### PR TITLE
Add HighDpiMode property/ApplyHighDpiMode Event to WindowsFormsApplicationBase.

### LINKED ISSUE #5069: Introduce ApplyApplicationDefaults Event to the Application Framework
## Rationale

In the WinForms 3.1-time frame, we introduced the `SetHighDpiMode` method and the `HighDpiMode` property. The reason for that was that we wanted to change the default HighDpiMode for WinForms Apps to be SystemAware. At the same time, we wanted to provide also a programmatically way to set the DpiMode, since the only remaining option to do that was over the App’s manifest – the App.Config option we were no longer supposed to use for .NET Core Apps. However, setting/changing the HighDpiMode via the manifest requires the App to be recompiled/rolled out anew, which was the reason to introduce the API as an alternative. Since we did not want to support the Visual Basic Application Framework in the time, we did not include new API options for the AF.
We decided however to enable the Application Framework for Visual Basic WinForms App (somewhat late) in the .NET 5-time frame. Unfortunately, it was to late to change the runtime at the time to include some necessary changes for setting the HighDpiMode programmatically in the Application Framework, so we decided to provide this event (and the setting of the correct default value) via an additional template file for Application Framework Apps, which is only active in .NET 5 Apps. We decided, we wanted to implement those events for .NET 6, and then supersede the template implementation by the actual API.

Visual Basic Application Framework Apps ever have had a different way compared to C# to configure the start-up behavior of an WinForms app: While we have _Program.cs_ in C#, there is no equivalent in VB with the Application Framework active. Rather, typical startup behavior is configured over several Application events:

For this reason, setting the HighDpiMode shouldn’t have been an exception conceptionally, so we introduced the `ApplyHighDpiMode` Application event via the template.

Since then, WinForms also has the necessity to set the default Font. The reason for that is, that we changed the default Font in .NET 3.1. That had an impact on pixel perfect Form designs, which were based on the previous Font settings. To accommodate User request for migration projects not needing to redesign all the Forms which based on the original Font, we decided to introduce the `SetDefaultFont` method for .NET 6. Again, this setting needs to get applied to the Application object, and since – again – in an AF App there is no _Program.cs_, we want to introduce an additional Event to set this default value. However, in contrast to C#, we need to make sure not to spam the Event list with yet additional Application events, which is why we decided to take a more holistic approach for AF by introducing the `ApplyApplicationDefaults` event for the AF. Now, with it, we will introduce a few new Properties to set as application wide default values, but unfortunately also a breaking change: We want to phase out `ApplyHighDpiMode`!

So, we will have in the first go of the Default values:
* Font - will be setting the DefaultFont.
* HighDpiMode - will be setting the HighDpiMode and replace the template-implemented ApplyHighDpiMode Application event. This is _only_ a breaking change, if folks have actually set this value via the Application Events.
* MinimumSplashScreenDisplayTime - it's a valuable property to set, but could previously only be set by overwriting `OnInitialize` of the `WindowsFormsApplicationBase` class. Customers shouldn't have such a hard time to discover how to set this value.

Since we’re sure to add more default properties in the future, this EventArgs class can easily be extended to accommodate more default settings (like `Margin`, `Padding`, default values for right-to-left settings, Default Form settings and more.) For the methodology to set default values, we decided that we do _not_ need to be in sync with C#, and the VB Application Framework can have its own identity.

### Actual APIs to be added:

This comes together in `WindowsFormsApplicationBase`, which will be extended by this:
```
        Public Event ApplyApplicationDefaults As ApplyApplicationDefaultsEventHandler

        ''' <summary>
        ''' We need this, so Project System will have a chance to set this value 
        ''' when we will implement this via a combo box for the VB Application Framework Designer.
        ''' </summary>
        Protected Property HighDpiMode() As Windows.Forms.HighDpiMode

```

For that, we need a new `EventArgs` class:

```
Public Class ApplyApplicationDefaultsEventArgs
    Inherits EventArgs

    Public Property DefaultFont As Font
    Public Property HighDpiMode As System.Windows.Forms.HighDpiMode
    Public Property MinimumSplashScreenDisplayTime As Integer
End Class
```

And finally, we also need this delegate for the event:

```
    ''' <summary>
    ''' Signature for the ApplyApplicationDefaults event handler
    ''' </summary>
    <EditorBrowsable(EditorBrowsableState.Advanced)>
    Public Delegate Sub ApplyApplicationDefaultsEventHandler(sender As Object, e As ApplyDefaultsEventArgs)
```

[UPDATE 7/10/2021: We will get rid of the `ApplyHighDpiSetting` event and unify that with the `ApplyDefaultValues` event by having the `HighDpiMode` property on the `ApplyApplicationDefaultsEventArgs`. We also introduce the `MinimumSplashScreenDisplayTime` property.]

### PR DESCRIPTION
**Pull request by KlausLoeffelmann**
This PR fixes #5069.

### CODE REVIEW

File: src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/WindowsFormsApplicationBase.vb [hunk 1]
```diff
@@ -680,6 +635,18 @@ Namespace Microsoft.VisualBasic.ApplicationServices
             End Set
         End Property
 
+        ''' <summary>
+        ''' Sets the HighDpiMode for the Application.
```
Comment (RussKie) [inline src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/WindowsFormsApplicationBase.vb:639]: I believe get/set properties are annotated as "Gets or sets..." ([example](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.button.autosizemode?view=netcore-3.1))
```suggestion
        ''' Gets or sets the HighDpiMode for the application.
```

File: src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/CantStartSingleInstanceException.vb [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+Imports Microsoft.VisualBasic.CompilerServices.Utils
+
+Namespace Microsoft.VisualBasic.ApplicationServices
+
+    ''' <summary>
+    ''' Exception for when we launch a single-instance application and it can't connect with the 
+    ''' original instance.
+    ''' </summary>
+    <EditorBrowsable(EditorBrowsableState.Never)>
+    <Serializable()>
```
[the remark this answers is not in the document]
Comment (KlausLoeffelmann) [reply src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/CantStartSingleInstanceException.vb:19]: >I don't believe exceptions are serializable any longer.

I just cleaned up the code but did not change anything except the lines I pointed out in the description.
I am hesitant to change this in this context, so let's leave this for now?

File: src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/StartupEventArgs.vb [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+Imports System.Collections.ObjectModel
+Imports System.ComponentModel
+Imports System.Runtime.InteropServices
+
+Namespace Microsoft.VisualBasic.ApplicationServices
+
+    ''' <summary>
+    ''' Provides context for the Startup event.
+    ''' </summary>
+    <EditorBrowsable(EditorBrowsableState.Advanced), ComVisible(False)>
```
[the remark this answers is not in the document]
Comment (KlausLoeffelmann) [reply src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/StartupEventArgs.vb:18]: I would suggest a dedicated cleanup where we'd change code for the VB part in a additional PR, but not in this context?

File: src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/WindowsFormsApplicationBase.vb (hunk 1 above)
Comment (KlausLoeffelmann) [reply src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/WindowsFormsApplicationBase.vb:639]: Sure!

**Review by RussKie — APPROVED** [review]

File: src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/CantStartSingleInstanceException.vb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+        Public Sub New(ByVal message As String)
+            MyBase.New(message)
+        End Sub
+
+        Public Sub New(ByVal message As String, ByVal inner As System.Exception)
+            MyBase.New(message, inner)
+        End Sub
+
+        ' Deserialization constructor must be defined since we are serializable
```
Comment (Tanya-Solyanik) [inline src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/CantStartSingleInstanceException.vb:37]: For other serializable types we added comments with descriptions of serialization scenarios, could you please describe serialization scenario here as well?

Comment (KlausLoeffelmann) [reply src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/CantStartSingleInstanceException.vb:37]: At this time, I will not address other things than HighDpiMode.
I will add a clean-up task for VB code in the 16.9/NET6 timeframe.

File: src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/ApplyApplicationDefaultsEventArgs.vb [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    ''' </summary>
+    <EditorBrowsable(EditorBrowsableState.Advanced), ComVisible(False)>
+    Public Class ApplyApplicationDefaultsEventArgs
+        Inherits EventArgs
+
+        ''' <summary>
+        ''' Creates a new instance of the StartupEventArgs.
+        ''' </summary>
+        Public Sub New(applicationDefaults As ApplicationDefaults)
+            _ApplicationDefaults = applicationDefaults
```
Comment (KlausLoeffelmann) [inline src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/ApplyApplicationDefaultsEventArgs.vb:26]: `_ApplicationDefaults` is the compiler-generated backing field, btw.

File: src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/CantStartSingleInstanceException.vb [hunk 6]
```diff
@@ -0,0 +1,43 @@
+﻿' Licensed to the .NET Foundation under one or more agreements.
```
Comment (KlausLoeffelmann) [inline src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/CantStartSingleInstanceException.vb:1]: This class is not new, just placed in a dedicated code file.

File: src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/CantStartSingleInstanceException.vb [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+Namespace Microsoft.VisualBasic.ApplicationServices
+
+    ''' <summary>
+    ''' Exception for when we launch a single-instance application and it can't connect with the 
+    ''' original instance.
+    ''' </summary>
+    <EditorBrowsable(EditorBrowsableState.Never)>
+    <Serializable()>
+    Public Class CantStartSingleInstanceException : Inherits System.Exception
```
Comment (KlausLoeffelmann) [inline src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/CantStartSingleInstanceException.vb:20]: This class is not new, just placed in a dedicated code file.

File: src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/NoStartupFormException.vb [hunk 8]
```diff
@@ -0,0 +1,42 @@
+﻿' Licensed to the .NET Foundation under one or more agreements.
```
Comment (KlausLoeffelmann) [inline src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/NoStartupFormException.vb:1]: This class is not new, just placed in a dedicated code file.

File: src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/StartupEventArgs.vb [hunk 9]
```diff
@@ -0,0 +1,38 @@
+﻿' Licensed to the .NET Foundation under one or more agreements.
```
Comment (KlausLoeffelmann) [inline src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/StartupEventArgs.vb:1]: This class is not new, just placed in a dedicated code file.

File: src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/StartupEventArgs.vb [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+Imports System.ComponentModel
+Imports System.Runtime.InteropServices
+
+Namespace Microsoft.VisualBasic.ApplicationServices
+
+    ''' <summary>
+    ''' Provides context for the Startup event.
+    ''' </summary>
+    <EditorBrowsable(EditorBrowsableState.Advanced), ComVisible(False)>
+    Public Class StartupEventArgs
```
Comment (KlausLoeffelmann) [inline src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/StartupEventArgs.vb:19]: This class is not new, just placed in a dedicated code file.

File: src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/StartupNextInstanceEventArgs.vb [hunk 11]
```diff
@@ -0,0 +1,46 @@
+﻿' Licensed to the .NET Foundation under one or more agreements.
```
Comment (KlausLoeffelmann) [inline src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/StartupNextInstanceEventArgs.vb:1]: This class is not new, just placed in a dedicated code file.

File: src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/UnhandledExceptionEventArgs.vb [hunk 12]
```diff
@@ -0,0 +1,32 @@
+﻿' Licensed to the .NET Foundation under one or more agreements.
```
Comment (KlausLoeffelmann) [inline src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/UnhandledExceptionEventArgs.vb:1]: This class is not new, just placed in a dedicated code file.

File: src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/WindowsFormsApplicationBase.WinFormsAppContext.vb [hunk 13]
```diff
@@ -0,0 +1,57 @@
+﻿Option Strict On
```
Comment (KlausLoeffelmann) [inline src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/WindowsFormsApplicationBase.WinFormsAppContext.vb:1]: This class is not new, just placed in a dedicated code file.

File: src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/CantStartSingleInstanceException.vb (hunk 6 above)
Comment (RussKie) [reply src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/CantStartSingleInstanceException.vb:1]: Let's move these out in a separate PR (such change can be merged straight away) and reduce the code churn and review overhead here

Comment (KlausLoeffelmann) [reply src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/CantStartSingleInstanceException.vb:1]: No, please, they really get in the way, and the 'Move-to-separate-file' function is terribly broken in VB, so this was quite an effort. Yes, I should have done this probably separately, but it is a straight forward change, and I have marked those.

File: src/System.Windows.Forms/src/System/Windows/Forms/Control.cs [hunk 14]
```diff
@@ -11614,7 +11614,9 @@ private void UserPreferenceChanged(object sender, UserPreferenceChangedEventArgs
             if (pref.Category == UserPreferenceCategory.Color)
             {
                 s_defaultFont = null;
-                Application.ScaleDefaultFont();
+
+                // TODO: IGOR: Why do we need to reset the Font when the Color has changed?
```
Comment (KlausLoeffelmann) [inline src/System.Windows.Forms/src/System/Windows/Forms/Control.cs:11618]: @RussKie, could you take a look here?

**Comment by RussKie** [comment]
Please rebase and clean up the history of the PR, there are multiple unrelated and non-functional changes that make it very difficult to reason about the real change in a combined diff and it is impossible to see those in a diff-by-diff scenario.

I don't quite understand why it is difficult to extract the non-functional changes (i.e. like types moving into own files) into own PR, but at least move these changes in a dedicated commit that is clearly labelled as such.

**Comment by KlausLoeffelmann** [comment]
Yes, you're right.
Will do.

File: src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/UnhandledExceptionEventArgs.vb [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+Imports System.ComponentModel
+Imports System.Runtime.InteropServices
+Imports System.Threading
+
+Namespace Microsoft.VisualBasic.ApplicationServices
+
+    ''' <summary>
+    ''' Provides the exception encountered along with a flag on whether to abort the program
+    ''' </summary>
+    <EditorBrowsable(EditorBrowsableState.Advanced), ComVisible(False)>
```
[the remark this answers is not in the document]
Comment (KlausLoeffelmann) [reply src/Microsoft.VisualBasic.Forms/src/Microsoft/VisualBasic/ApplicationServices/UnhandledExceptionEventArgs.vb]: I would like to do it in a later, dedicated step. Leaving this for now.

**Review by JeremyKuhne — APPROVED** [review]

**Review by merriemcgaw — APPROVED** [review]

**Review by JeremyKuhne — APPROVED** [review]

**Comment by RussKie** [comment]
❗ Squash merge unless there is a significant reason to preserve history.

**Comment by KlausLoeffelmann** [comment]
Always? 
The agreed default in the Designer is exactly the opposite, btw.

**Comment by RussKie** [comment]
Yes, always. Not doing so pollutes the history is unnecessary commits (no one really cares how many time an author fixed a typo or addressed feedback in the PR), and it makes it
* harder to reason about the history, 
* significantly more difficult to investigate any issues introduced by the PR (need to do multiple blame passes), and 
* extremely difficult to cherry-pick (if necessary).

**Pull request merged**